### PR TITLE
Allow loading LLVM plugins with both legacy and new pass manager 

### PR DESCRIPTION
**Pull request by eskarn**
Opening a draft PR to get feedback and start discussion on this feature. There is already a codegen option `passes` which allow giving a list of LLVM pass names, however we currently can't use a LLVM pass plugin (as described here : https://llvm.org/docs/WritingAnLLVMPass.html), the only available passes are the LLVM built-in ones.

The proposed modification would be to add another codegen option `pass-plugins`, which can be set with a list of paths to shared library files. These libraries are loaded using the LLVM function `PassPlugin::Load`, which calls the expected symbol `lvmGetPassPluginInfo`, and register the pipeline parsing and optimization callbacks.

An example usage with a single plugin and 3 passes would look like this in the `.cargo/config`:

```toml
rustflags = [
    "-C", "pass-plugins=/tmp/libLLVMPassPlugin",
    "-C", "passes=pass1 pass2 pass3",
]
```
This would give the same functionality as the opt LLVM tool directly integrated in rust build system.

Additionally, we can also not specify the `passes` option, and use a plugin which inserts passes in the optimization pipeline, as one could do using clang.

### CODE REVIEW

**Comment by nagisa** [comment]
How is this different from the flag introduced by https://github.com/rust-lang/rust/pull/86267?

**Comment by eskarn** [comment]
> How is this different from the flag introduced by #86267?

The main difference is using the new LLVM pass manager registration, which is not possible with the current flag.

The flag `llvm-plugin` allows loading an arbitrary library, and then if the library happens to use the legacy pass manager registration, then its passes can be registered.

The proposed flag would allow loading only libraries which use the new LLVM pass manager format, which means those exposing the symbol `lvmGetPassPluginInfo`, which gets loaded by `PassPlugin::Load`.
There is also verification of the `LLVM_PLUGIN_API_VERSION`, which is declared during the new PM registration.

**Comment by nagisa** [comment]
Can we make the already existing flag work for both oldPM and newPM plugins?

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -995,6 +995,8 @@ options! {
         "panic strategy to compile crate with"),
     passes: Vec<String> = (Vec::new(), parse_list, [TRACKED],
         "a list of extra LLVM passes to run (space separated)"),
+    pass_plugins: Vec<String> = (Vec::new(), parse_list, [TRACKED],
+        "a list of LLVM pass plugins to load (space separated)"),
```
[the remark this answers is not in the document]
Comment (eskarn) [reply compiler/rustc_session/src/options.rs:999]: > This needs to be under debugging options as it should be an unstable option. Also please prefix it with `llvm_` as it is LLVM specific.

Ok thanks, seems easy to change this. I've tried to stay close to the existing `passes` option, but maybe it should also have been prefixed and a debugging option then.

**Comment by eskarn** [comment]
> 

`PassPlugin::Load` won't load a library without the expected symbol, so we'd need to handle the legacy plugins separately. Keeping only the existing `llvm-plugins` option, something like this could work:
- if the option `new_llvm_pass_manager` is true, then we skip this code entirely :
https://github.com/rust-lang/rust/blob/22c2d9ddbf356bcdb718e88ca6ee3665e1e42690/compiler/rustc_codegen_llvm/src/llvm_util.rs#L123-L131
Then we just keep the proposed modifications in compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (only using the existing `llvm-plugins` option instead).
- if the option `new_llvm_pass_manager` is false, we run the code in llvm_util.rs and `LLVMRustOptimizeWithNewPassManager` won't be called (so we won't try to load the plugins as new PM ones)

The only issue would be when the option isn't specified and the arch is `s390x`, there is an exception here: https://github.com/rust-lang/rust/blob/8b09ba6a5d5c644fe0f1c27c7f9c80b334241707/compiler/rustc_codegen_llvm/src/back/write.rs#L409
For this specific case, the user of the `llvm-plugins` would have to know the new PM plugins are not usable because of the exception.

So basically, to use the new PM plugins we'd have to actively set `new_llvm_pass_manager` to true (and not use the `s390x` arch).

Would this solution seem acceptable ?

**Comment by nagisa** [comment]
Making it conditional on a flag seems reasonable (oldPM is going away shortly anyway), though I was kind of hoping that `PassPlugin::Load` would handle plugins for either PM.

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 2]
```diff
@@ -118,14 +118,20 @@ unsafe fn configure_llvm(sess: &Session) {
...
-        let path = Path::new(plugin);
-        let res = DynamicLibrary::open(path);
-        match res {
-            Ok(_) => debug!("LLVM plugin loaded succesfully {} ({})", path.display(), plugin),
-            Err(e) => bug!("couldn't load plugin: {}", e),
+    let use_new_llvm_pm_plugin_register =
+        sess.opts.debugging_opts.new_llvm_pass_manager.unwrap_or(false);
+
+    // Use the legacy pm registration if the new_llvm_pass_manager option isn't explicitly enabled
+    if use_new_llvm_pm_plugin_register {
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:125]: Should this condition be inverted?

Comment (eskarn) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs]: Indeed, it should be fixed with 6e703ce333f193607329172a1f4600c4ff51227c.

**Comment by eskarn** [comment]
With 0ce67ab389746cd4e135ed15b3c2eb7c32c0d064 and 0ce67ab389746cd4e135ed15b3c2eb7c32c0d064, now we only load new PM plugins from `llvm-plugins` if `new-llvm-pass-manager` is set to yes explicitly. Otherwise the behavior is the same as before, loading legacy plugins only.

These changes should address current @bjorn3 and @nagisa feedback.

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 3]
```diff
@@ -118,14 +118,20 @@ unsafe fn configure_llvm(sess: &Session) {
...
     llvm::LLVMInitializePasses();
 
-    for plugin in &sess.opts.debugging_opts.llvm_plugins {
-        let path = Path::new(plugin);
-        let res = DynamicLibrary::open(path);
-        match res {
-            Ok(_) => debug!("LLVM plugin loaded succesfully {} ({})", path.display(), plugin),
-            Err(e) => bug!("couldn't load plugin: {}", e),
+    let use_new_llvm_pm_plugin_register =
+        sess.opts.debugging_opts.new_llvm_pass_manager.unwrap_or(false);
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:122]: If `new_llvm_pass_manager` returns `None`, the new pass manager is used when the llvm version is at least 13 and the target is not s390x. See https://github.com/rust-lang/rust/blob/8b09ba6a5d5c644fe0f1c27c7f9c80b334241707/compiler/rustc_codegen_llvm/src/back/write.rs#L397-L410

Comment (eskarn) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:122]: Yes that's what i was mentioning at the end of https://github.com/rust-lang/rust/pull/91125#issuecomment-976710806.

What i'm proposing is to condition the new PM plugin loading only on the explicit `new_llvm_pass_manager=yes` and not the actual usage of the new PM later. Since there are exceptions and a default value hardcoded, we can't trust the option to reflect reality.

If we do want to condition on the actual usage of the new PM then i see two potential issues:
- current users of `llvm-plugins` when updating to a llvm13 based rust will trigger PassPlugin::Load on their legacy plugins since the new PM is enabled by default. They will need to explicitly set `new_llvm_pass_manager=false`
- i'm not sure how to use `should_use_new_llvm_pass_manager` in llvm_util.rs, maybe just move it to llvm_util.rs to make it more widely available ?

**Comment by eskarn** [comment]
Resolved conflicts with #90716, re-tested with upstream, and updated the PR title to reflect more what is now being proposed.

File: compiler/rustc_codegen_llvm/src/llvm_util.rs (hunk 3 above)
Comment (eskarn) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:123]: @bjorn3 do you think the current code is acceptable regarding the explicit usage of `new_llvm_pass_manager` or i should
propose the alternative described in my last reply ?

Comment (bjorn3) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:123]: Sorry for the late reply. Missed your previous comment.

> current users of llvm-plugins when updating to a llvm13 based rust will trigger PassPlugin::Load on their legacy plugins since the new PM is enabled by default. They will need to explicitly set new_llvm_pass_manager=false

At least in that case you will get an error message rather than silently ignoring it, right?

> i'm not sure how to use should_use_new_llvm_pass_manager in llvm_util.rs, maybe just move it to llvm_util.rs to make it more widely available ?

Makes sense I think.

> do you think the current code is acceptable regarding the explicit usage of new_llvm_pass_manager or i should
propose the alternative described in my last reply ?

I think using `should_use_new_llvm_pass_manager` is better. I don't strongly object to the current code though.

Comment (eskarn) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:123]: > Sorry for the late reply. Missed your previous comment.
> 
No problem, i should have pinged you earlier but i was busy on other tasks anyway.
> > current users of llvm-plugins when updating to a llvm13 based rust will trigger PassPlugin::Load on their legacy plugins since the new PM is enabled by default. They will need to explicitly set new_llvm_pass_manager=false
> 
> At least in that case you will get an error message rather than silently ignoring it, right?
> 
Indeed, in that case  `cargo build` fails with an error and i can see original LLVM error message directly in cargo non verbose logs like this: `Plugin entry point not found in '/tmp/libLLVMPlugin.so'. Is this a legacy plugin?`
> > i'm not sure how to use should_use_new_llvm_pass_manager in llvm_util.rs, maybe just move it to llvm_util.rs to make it more widely available ?
> 
> Makes sense I think.
Ok let's try it.
> 
> > do you think the current code is acceptable regarding the explicit usage of new_llvm_pass_manager or i should
> > propose the alternative described in my last reply ?
> 
> I think using `should_use_new_llvm_pass_manager` is better. I don't strongly object to the current code though.
Ok thank you for your feedback, i'll try doing it this way and see how that goes.

**Comment by nagisa** [comment]
Please click the “ready for review” button once you think its good to go.
